Add custom formatting to IRC & Discord output

Throne3d opened the pull request:
As requested by #13, it might sometimes be useful to have custom formatting in the output for IRC and Discord.

I'm not sure how exactly to add tests for this, as it seems like it's going to need a separate configuration for (almost) each test, and this doesn't look like it'll fit into the current `bot.test.js` file very easily. From a brief glance I can't see a 'context' feature in Chai as there is in RSpec, which would allow me to limit the `beforeEach` to just the tests present, and add separate ones at the end for this. It's possible I'm missing something.

I'm also not too strongly attached to this idea and it might make it harder to implement some things in the future, so if it'd be preferable to drop this, I don't care too strongly. (If people want it desperately enough, they can always just patch these commits on – hopefully it's clearly documenting enough, and it doesn't seem to break current tests.)

I'd appreciate feedback on how to proceed. Thanks.

Throne3d commented:
I guess I could add a new "describe" context, within the current one, for the tests that use a regular-config bot, and separate out the `beforeEach` into two bits? (Then use another context for new tests that don't.)

Throne3d commented:
I've added tests for this (which should hopefully now pass, making this not a WIP), but perhaps instead of creating a new `describe` block to separate out the default-config bot tests, I should just do as some tests have already and redefine `bot` with a different config in the tests that need it.

Opinions?

Throne3d commented:
Nm, undid the contexting and just manually used `const bot` in the appropriate tests.

ekmartin commented on lib/bot.js:
```diff
@@ -148,6 +168,13 @@ class Bot {
 
+  static substitutePattern(message, patternMapping) {
+    return message.replace(patternMatch, (match, varName) => {
+      if (varName in patternMapping) return patternMapping[varName];
```
Could simplify this to `(match, varName) => patternMapping[varName] || match`?

Throne3d replied:
The main reason I didn't was in case the patternMapping was a blank string. I guess this probably isn't going to happen anywhere (unless there was somehow a zero-length discord message? But there's a guard for that), so it should be fine to simplify.

ekmartin commented on lib/bot.js:
```diff
@@ -33,6 +34,25 @@ class Bot {
+    this.formatCommandPrelude = options.formatCommandPrelude || 'Command sent from Discord by {$nickname}:';
+    this.formatIRCText = options.formatIRCText || '<{$displayUsername}> {$text}';
+    this.formatURLAttachment = options.formatURLAttachment || '<{$displayUsername}> {$attachmentURL}';
+    // "{$keyName}" => "variableValue"
```
Nitpick, but I think it makes more sense to put these comments above the variables.

Throne3d replied:
Makes sense.

ekmartin commented on lib/bot.js:
```diff
@@ -33,6 +34,25 @@ class Bot {
     this.ircNickColor = options.ircNickColor !== false; // default to true
     this.channels = _.values(options.channelMapping);
 
+    this.formatCommandPrelude = options.formatCommandPrelude || 'Command sent from Discord by {$nickname}:';
```
Maybe create a `format` object and have these as properties? I.e.
```js
"format": {
  "commandPrelude": "...",
  "ircText": "...",
  "urlAttachment": "..."
}
```

or something like that? Then just add a `this.format = options.format || {};` above?

ekmartin commented:
Looks great, thanks! I've normally been against adding too many configuration options, but I think this could be useful.

Throne3d commented:
(Rebased onto master, rebased to move comments, updated to work with the new formatting-bridging thing (`{$text}` is now relevantly formatted), added a commit to move the formatting configs to a `format` object, added a commit to simplify substitutePattern.)

ekmartin commented:
Released in [2.3.0](https://github.com/reactiflux/discord-irc/blob/master/CHANGELOG.md#230---2017-04-03)!
